File: 3-longest-substring-without-repeating-characters/3-longest-substring-without-repeating-characters.js
/**
 * @param {string} s
 * @return {number}
 */

/*
ask: what's the longest substring that ends at an index?

data structures:
substring to keep track of longest substring ending at an index
array to keep track of length of longest substring ending at an index
*/

var lengthOfLongestSubstring = function(s) {
    if (!s.length) return 0;
    let sub = '';
    let max = 0;
    
    for (let i = 0; i < s.length; i++) {
        if (!sub.includes(s[i])) {
            sub += s[i];
        } else {
            sub = sub.slice(sub.indexOf(s[i]) + 1) + s[i];
        }
        if (sub.length > max) max = sub.length;
    }
    
    return max;
};

File: 3-longest-substring-without-repeating-characters/3-longest-substring-without-repeating-characters.py
# data structures:
# curr_sub to keep track of current longest substring that ends on index

class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        if s == '':
            return 0
        curr_sub, maxCount = '', 0
        
        for i in range(len(s)):
            if s[i] in curr_sub:
                curr_sub = curr_sub[curr_sub.index(s[i])+1:] + s[i]
            else:
                curr_sub += s[i]
            if len(curr_sub) > maxCount:
                maxCount = len(curr_sub)
            
        return maxCount